src/java.base/share/classes/java/lang/StringFormat.java

```diff
@@ -0,0 +1,327 @@
+ * questions.
+ */
+
+package java.lang;
+
+import jdk.internal.math.DoubleToDecimal;
+import jdk.internal.util.HexDigits;
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.Formatter;
+import java.util.IllegalFormatConversionException;
+import java.util.function.LongToIntFunction;
+import java.util.function.ToIntFunction;
+
+/**
+ * Utility class for string format fastpath
+ */
+final class StringFormat {
+    static final char DECIMAL_INTEGER           = 'd';
+    static final char HEXADECIMAL_INTEGER       = 'x';
+    static final char HEXADECIMAL_INTEGER_UPPER = 'X';
+    static final char STRING                    = 's';
+
+    static String format(String format, Object... args) {
+        if (args != null) {
+            String s = null;
+            if (args.length == 1) {
+                s = format1(format, args[0]);
+            } else if (args.length == 2) {
+                s = format2(format, args[0], args[1]);
+            }
+            if (s != null) {
+                return s;
+            }
+        }
+
+        return new Formatter().format(format, args).toString();
+    }
+
+    private static String format1(String format, Object arg) {
+        int max = format.length();
+        int off = format.indexOf('%');
+        if (off == -1 || off + 1 == max) {
+            return null;
+        }
+
+        if (off + 2 < max) {
+            if (format.indexOf('%', off + 2) != -1) {
+                return null;
+            }
+        }
+
+        char conv = format.charAt(off + 1);
+        int width = 0;
+        if (conv >= '1' && conv <= '9') {
+            width = conv - '0';
+            if (off + 2 < max) {
+                conv = format.charAt(off + 2);
+            }
+        }
+
+        if (conv == STRING && isBigInt(arg)) {
+            conv = DECIMAL_INTEGER;
+        }
+
+        if (conv == STRING) {
+            arg = String.valueOf(arg);
+        }
+
+        int size = stringSize(conv, arg);
+        if (size == -1) {
+            return null;
+        }
+        return format1(format, off, conv, arg, width, size);
+    }
+
+    private static String format2(String format, Object arg0, Object arg1) {
+        int max = format.length();
+        int off0 = format.indexOf('%');
+        if (off0 == -1 || off0 + 1 == max) {
+            return null;
+        }
+        char conv0 = format.charAt(off0 + 1);
+        int width0 = 0;
+        if (conv0 >= '1' && conv0 <= '9') {
+            width0 = conv0 - '0';
+            if (off0 + 2 < max) {
+                conv0 = format.charAt(off0 + 2);
+            }
+        }
+
+        int off1 = format.indexOf('%', off0 + 1);
+        if (off1 == -1 || off1 + 1 == max) {
+            return null;
+        }
+        char conv1 = format.charAt(off1 + 1);
+        int width1 = 0;
+        if (conv1 >= '1' && conv1 <= '9') {
+            width1 = conv1 - '0';
+            if (off1 + 2 < max) {
+                conv1 = format.charAt(off1 + 2);
```

Comment:
is it worth handling `width1 > 9`?

```suggestion
        for (int i = 2; conv1 >= '1' && conv1 <= '9' && off1 + i < max; i++) {
            width1 = (width1 * 10) + (conv1 - '0');
            if (off1 + i < max) {
                conv1 = format.charAt(off1 + i);
```